ls1intum/Artemis
# [Feature] Allow structured grading instructions editable after first usage

### Checklist

### Motivation and Context


The structured grading instructions enable instructors and teaching assistants to provide standard feedback during the assessment. Artemis does not support to change structured grading instructions when the feedback of instructions is used in the assessment. In some cases, misspelling in the instructions required to be fixed during the assessment. Also, Instructors/TAs can not add new grading instruction during the assessment. As a result, the instructors and the teaching assistants have to write feedback manually and the instructions are not very useful for them. This PR provides a solution for the instructors, and they can edit the instructions in a more structured way.

### Description


Part I
To understand if the structured grading instruction is used or not used, `isGradingInstructionFeedbackUsed` is set for the exercise when the server sends exercise details to client side. If it is used, the grading instructions are displayed in form (similar with the re-evaluation of quiz exercises) 
`Feedback `and `GradingInstruction `have many-to-one relationship, so deleting only grading instruction would lead data integrity violation. Instructors could want that all corresponding feedback is deleted, but they could also want that only the association is removed and the feedback still stays valid. Our solution is when the grading instruction deleted, the association between `GradingInstruction `and `Feedback `is removed. The follow-up task could be some option to remove the all feedback which is associated with the grading instruction that will be deleted.

Part II (This will be solved in a follow up PR. Until then, user will be warned when he/she change the instruction credit or delete the instruction)
If the existing instruction which is associated with `Feedback` is updated by Instructor, the corresponding feedback should also updated. Re-evaluation of feedback is not implemented yet, it requires additional tasks. When it is done, the feedback text, credit would be updated and the result would be re-assessed. While the user is updating the exercise, Artemis would warn the instructor/editor that these changes have consequences for the existing assessments. 

### Steps for Testing


1. Log in to Artemis
2. Create/Find a new exercise (could be any type of exercise except quiz)
3. Switch structured grading instruction, then save the exercise.
4. Add an example solution for the exercise / participate this exercise as a student
5. If you add example submission, then create new assessment for this submission and use one of the structured grading instruction as a feedback. (You can drag&drop the instruction into the feedback area. see >> screen record)
5. If you participated this exercise as student, then login as tutor/instructor and assess this submission. During the assessment, use one of the structured grading instructions as a feedback. (You can drag&drop the instruction into the feedback area. see >> screen record)
6. After that, as an instructor, go to edit view for this exercise.
7. switch the structured grading instruction button, after first usage of the feedback, the instructions should be displayed in form.
8. You can add/edit/delete grading criteria title or grading instructions. please test the all new buttons.
9. Then try to save the exercise, there will be no "Internal Server Error", and you can see updated grading instruction in the exercise detail page.
    a) if you delete or change the credit for the existing instructions, the warning pop-up should be displayed after you click "Save" button. 
       (see screenshot) The exercise should be updated when you click "Confirm Changes".



**You should also check that structured grading instructions can be edited when there is no submissions which means that grading instructions are not used as feedback.

***If you use grading instructions to asesss modeling exercise UML elements inside of the Apollon Editor (see last screenshot), currently it can not match the instruction with the feedback. So, wehen you go back to edit page, you can not see the grading instructions in a new form. It requires additional information from Apollon editor. This will be solved in a follow up.

### Test Coverage




grading-instructions-details.component.ts 59% (only new functions are  covered, will be increased in the follow-up task)
FileUploadExerciseResource.java 100%
TextExerciseResource.java 92%
ModelingExerciseResource.java 93%

### Screenshots



Following view represents the current version of structured grading instruction.

<img src="https://user-images.githubusercontent.com/23650630/116827748-f5146580-ab9a-11eb-8314-38db26fc6182.png" width="400">

Following view represents the  structured grading instruction if one of them used as a feedback.

<img src="https://user-images.githubusercontent.com/23650630/116827802-4ae90d80-ab9b-11eb-8c94-6c6897775e26.png" width="400">

How to use structured grading instruction as feedback during the assessment period

https://user-images.githubusercontent.com/23650630/116828319-3c502580-ab9e-11eb-84ea-b48414edcbb1.mp4

The warning pop-up only if you change the instruction credit or delete one of the instruction/criteria which is associated with the feedback

<img src="https://user-images.githubusercontent.com/23650630/117378695-34fa8600-aed6-11eb-9af4-c8761b6c0e35.png" width="400">

Modeling Assessment for UML elements - user can drag&drop the single instruction into the UML element

<img src="https://user-images.githubusercontent.com/23650630/118262104-b47aed00-b4b4-11eb-9911-01f0de5b5108.png" width="400">
